Docs/unr 1133 singleton guidelines (#877)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/singleton-actors.md

<%(TOC)%>
# Singleton Actors

You can use a Singleton Actor to define a single source of truth for operations or data, or both, across a game world that uses zoning. You can have only one instance of an [entity]({{urlRoot}}/content/glossary#entity) that represents a Singleton Actor per game world.

You define a Singleton Actor by tagging an Actor with the `SpatialType=Singleton` class attribute. For example, if you are implementing a scoreboard, you probably want only one scoreboard in your world, so you can tag the scoreboard Actor as a Singleton Actor.

> There are two types of Singleton Actors:

Reviewer: Looking at this again, I think it should be "Actor" (singular). 

It's similar to "There are many types of dog" (you wouldn't say "many types of dogs").

Reviewer 2: That's an interesting point! 

I googled it and found there had been much discussion on that: 
- https://english.stackexchange.com/questions/29843/types-of-followed-by-singular-or-plural
- https://ffeathers.wordpress.com/2013/09/28/do-you-use-singular-or-plural-after-types-of/

In particular, I noticed that "... there's also variation between British and American English. I speak AmE and I don't think I would ever say there are two types of apple, as a BrE speaker in the linked question mentions is perfectly grammatical there. "

Reviewer: Oh wow, it looks like this is somewhat controversial and not clear-cut (like so much of English 🙄)

I've started a poll in the #terminology channel, to see if people have strong opinions... If the singular is a very UK-based thing, then we should avoid it, even though the plural sounds super strange to me!

Reviewer 2: I've just noticed from the #terminology channel "There are two Singleton Actor types", we can use that!

> * **Public Singleton Actors** - Singleton Actors that are replicated to [server-workers and client-worker instances]({{urlRoot}}/content/glossary#workers). [`AGameState`](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode) and `ALevelScriptActor` are Public Singleton Actors.

Reviewer: ```suggestion
* **Public Singleton Actors** - Singleton Actors that are replicated to [server-worker and client-worker instances]({{urlRoot}}/content/glossary#workers). [`AGameState`](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode) and `ALevelScriptActor` are Public Singleton Actors.
```

* **Private Singleton Actors** - Singleton Actors that are replicated to [server-worker instances]({{urlRoot}}/content/glossary#workers), but not accessible to [client-worker instances]({{urlRoot}}/content/glossary#workers). [`AGameMode`](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode) is a Private Singleton Actor.

You can define any class as a Singleton Actor. However, the following Unreal Engine classes are automatically tagged in source code as Singleton Actors, so you don’t need to define them:

* `AGameModeBase`
* `AGameStateBase`

> * `ALevelScriptActor`

Reviewer: Do we need a link for this one? I found https://api.unrealengine.com/INT/API/Runtime/Engine/Engine/ALevelScriptActor/index.html but I'm not sure if that's appropriate. It's not the same type of link as https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode.

Reviewer 2: Per Mike's reply, we can add a link to that though this link is not part of docs...

> > **Note**: Because the `SpatialType` class attribute is inheritable, all the classes that derive from `AGameModeBase`,  `AGameStateBase`, or `ALevelScriptActor` class are also Singleton Actors, which means that you don’t need to explicitly tag them in your code. You can opt out of this by tagging these classes with the [`NotSpatialType`]({{urlRoot}}/content/spatial-type#spatial-type) tag.

Reviewer: ```suggestion
> **Note**: Because the [`SpatialType`]({{urlRoot}}/content/spatial-type) class attribute is inheritable, all the classes that derive from `AGameModeBase`, `AGameStateBase`, or `ALevelScriptActor` are also Singleton Actors, which means that you don’t need to explicitly tag them in your code. You can opt out of this by tagging these classes with the [`NotSpatialType`]({{urlRoot}}/content/spatial-type#spatial-type) tag.
```

     ```
     UCLASS(SpatialType=(Singleton, ServerOnly))
     class AScoreBoard : public AActor
     {
       GENERATED_BODY()
       ...
     }
     ``` 

### Unreal Blueprint Class

1. Open your Blueprint Class in the [Blueprint Editor](https://docs.unrealengine.com/en-us/Engine/Blueprints/Editor).
2. From the [Toolbar](https://docs.unrealengine.com/en-US/Engine/Blueprints/Editor/UIComponents/Toolbar), click **Class Settings** to open Blueprint Properties in the Default pane.
3. In the **Class Options** section, ensure that the **Spatial Type** check box is selected.
4. In the **Spatial Description** text box, tag the Blueprint Class with the relevant attribute(s), depending on what type of Singleton Actor you want it to be:
   * **Public Singleton Actor**: enter `Singleton`.
   * **Private Singleton Actor**: enter `Singleton ServerOnly`.
  Here's a screenshot of what your Blueprint Class looks like if you tag it as a Private Singleton Actor:

![Singleton Blueprint]({{assetRoot}}assets/screen-grabs/blueprint-singleton.png)

> ## Managing Singleton Actors

Reviewer: I think you've missed out the section on spawning?

Reviewer 2: Thanks for pointing it out. I copied and pasted everything from the Gdocs here. When I tried to merge the changes, there were conflicts. It might occur while I was resolving the conflicts probably?

It's now added.

Because each server-worker instance spawns each Singleton Actor locally, we have introduced the [Global State Manager]({{urlRoot}}/content/glossary#global-state-manager) (GSM) to ensure that there is only ever one instance of an [entity]({{urlRoot}}/content/glossary#entity) that represents a Singleton Actor, and to ensure data is replicated properly. 

> The GSM allows only the server-worker instance with authority over the GSM to spawn a representative entity for each Singleton Actor. Every other server-worker instance spawns each Singleton Actor locally and then links it to the one on the server-worker instance that has authority over the GSM. Additionally, each Public Singleton Actor is replicated to every client-worker instance via the normal [Actor replication flow](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/ReplicationFlow).

Reviewer: ```suggestion
The GSM allows only the server-worker instance with [authority]({{urlRoot}}/content/glossary#authority) over the GSM to spawn a representative entity for each Singleton Actor. Every other server-worker instance spawns each Singleton Actor locally and then links it to the one on the server-worker instance that has authority over the GSM. Additionally, each Public Singleton Actor is replicated to every client-worker instance via the normal [Actor replication flow](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/ReplicationFlow).
```

Reviewer 2: I just doubt that users do want/need to know what's authority when they arrive here [with a certain level of knowledge].

Reviewer: It's a fair point, and I know we've discussed it before. It's possible that we tend to go over the top with our cross-linking in the docs!

3. In the **Class Options** section, ensure that the **Spatial Type** check box is selected.
4. In the **Spatial Description** text box, tag the Blueprint Class with the relevant attribute(s), depending on what type of Singleton Actor you want it to be:
   * **Public Singleton Actor**: enter `Singleton`.
   * **Private Singleton Actor**: enter `Singleton ServerOnly`.
  Here's a screenshot of what your Blueprint Class looks like if you tag it as a Private Singleton Actor:

![Singleton Blueprint]({{assetRoot}}assets/screen-grabs/blueprint-singleton.png)

## Spawning Singleton Actors

For all the following classes, which are automatically tagged as Singleton Actors, Singleton Actors are spawned automatically on all server-worker instances as part of the [Unreal Engine game flow](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameFlow):
- `AGameMode`
- `AGameState`
- `ALevelScriptActor`
- any classes that derive from the preceding classes

For all classes that you manually tag as Singleton Actors, make sure that Singleton Actors are spawned on all server-worker instances at runtime. You need to write code to do this when you create each server-worker type. 

> > *Note*: Never write code to spawn Singleton Actors when you create client-worker types.

Reviewer: ```suggestion
> **Note**: Never write code to spawn Singleton Actors when you create client-worker types.
```

<%(TOC)%>
# Singleton Actors

You can use a Singleton Actor to define a single source of truth for operations or data, or both, across a game world that uses zoning. You can have only one instance of an [entity]({{urlRoot}}/content/glossary#entity) that represents a Singleton Actor per game world.

You define a Singleton Actor by tagging an Actor with the `SpatialType=Singleton` class attribute. For example, if you are implementing a scoreboard, you probably want only one scoreboard in your world, so you can tag the scoreboard Actor as a Singleton Actor.

> There are two Singleton Actor types:

Reviewer: I think this is OK here. However, I don't particularly like it, because it makes me think that there might be a thing called a "Singleton Actor type". But it's not a big deal - I think people will understand what it means :)

Before you begin to use Singleton Actors across your game world, review the following high-level procedure. See each section for more details:

1. [Define Singleton Actors](#defining-singleton-actors): Manually tag an Unreal C++ class or an Unreal Blueprint class with the relevant attributes to be a Public Singleton Actor or Private Singleton Actor.
2. [Spawn Singleton Actors](#spawning-singleton-actors): Write code to spawn Singleton Actors on server-worker instances at runtime. 

In addition, learn about how SpatialOS can [manage Singleton Actors](#managing-singleton-actors) to ensure that there is only ever one instance of an entity that represents a Singleton Actor, and to ensure data is replicated properly. 

## Defining Singleton Actors

Before you begin, ensure that you have an Actor, which can be either an [Unreal C++ Class](https://docs.unrealengine.com/en-us/Programming/Development/ManagingGameCode/CppClassWizard) or an [Unreal Blueprint Class](https://docs.unrealengine.com/en-US/Engine/Blueprints/UserGuide/Types/ClassBlueprint).

### Unreal C++ Class

1. Open your C++ Class in any text editor.

> 2. Tag the C++ Class with the relevant attribute(s), depending on what type of Singleton Actor you want it to be:

Author: Should this number indexing all be `1`s?

Reviewer: Good point, yes it should! (@eddyrainy if you use "1" for every numbered step, this gets translated into the correct number in the actual docs. It means you can add more steps in the middle without having to re-number everything.)

### Unreal Blueprint Class

1. Open your Blueprint Class in the [Blueprint Editor](https://docs.unrealengine.com/en-us/Engine/Blueprints/Editor).
2. From the [Toolbar](https://docs.unrealengine.com/en-US/Engine/Blueprints/Editor/UIComponents/Toolbar), click **Class Settings** to open Blueprint Properties in the Default pane.
3. In the **Class Options** section, ensure that the **Spatial Type** check box is selected.
4. In the **Spatial Description** text box, tag the Blueprint Class with the relevant attribute(s), depending on what type of Singleton Actor you want it to be:
   * **Public Singleton Actor**: enter `Singleton`.
   * **Private Singleton Actor**: enter `Singleton ServerOnly`.
    
    Here's a screenshot of what your Blueprint Class looks like if you tag it as a Private Singleton Actor:
    ![Singleton Blueprint]({{assetRoot}}assets/screen-grabs/blueprint-singleton.png)

## Spawning Singleton Actors

For all the following classes, which are automatically tagged as Singleton Actors, Singleton Actors are spawned automatically on all server-worker instances as part of the [Unreal Engine game flow](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameFlow):

>   - `AGameMode`

Author: ```suggestion
  - `AGameModeBase`
```

>   - `AGameState`

Author: ```suggestion
  - `AGameStateBase`
```

  - `ALevelScriptActor`
  - any classes that derive from the preceding classes

> For all classes that you manually tag as Singleton Actors, make sure that Singleton Actors are spawned on all server-worker instances at runtime. You need to write code to do this when you create each server-worker type. 

Author: I'm not sure the second sentence is necessary. Spawning Singleton Actors needs to be done independently of the server-worker type, and we don't even official support other server-worker types currently.

Reviewer: Ah, I thought the spawning bit was some code you had to write as part of the code for setting up your worker type. But I don't have a good mental model of this, having never done it myself. How do you actually get (manually tagged) Singleton Actors to spawn on each worker instance?

Author: You'd spawn them like you would any normal actor. Most likely it would be part of the game mode's begin play logic. It's entirely up to the user. I think this is covered in the first sentence though.

Reviewer: OK, cool - as well as removing this sentence, would it make sense to replace

> **Note**: Never write code to spawn Singleton Actors when you create client-worker types.

with

> **Note**: Never spawn Singleton Actors on client-worker instances.

?

(To me, this sounds strange, because the user isn't directly doing the spawning - they're writing the code to do the spawning - but perhaps in context it makes sense?)

Author: Yes sounds good.

Reviewer 2: Since it's not *write code to worker types*, I think we also need to change the following sentence from:
> [Spawn Singleton Actors](#spawning-singleton-actors): Write code to spawn Singleton Actors on server-worker instances at runtime. 

possibly to

> [Spawn Singleton Actors](#spawning-singleton-actors): Understand how Singleton Actors are spawned on server-worker instances at runtime.

Reviewer: I think this update makes sense 👍

* As `SpatialType` is inheritable, all classes that derive off these classes are also considered Singleton Actors. You can opt out using the `NotSpatialType` tag.

# Singleton Spawning Rules

> 1. Never explicitly spawn any Singleton Actor on client-workers. The runtime will ensure each client receives the Singleton Actor via the normal actor replication pipeline.

Reviewer 3: ```suggestion
1. Never explicitly spawn any Singleton Actors on client-workers. The runtime will ensure each client receives the Singleton Actor via the normal actor replication pipeline.
```

> 1. Always explicity spawn all Singleton Actors on server-workers. The Unreal GDK will ensure that only one server-worker is authoritive over the Singleton Actor. See Coordinating Singletons.

Reviewer 3: ```suggestion
1. Always explicity spawn all Singleton Actors on server-workers. The Unreal GDK will ensure that only one server-worker is authoritive over the Singleton Actor. See [Coordinating Singletons](#coordinating-singletons).
```

> 1. A caveat to the above rule is `AGameMode` and `AGameState` Singletons Actors will be spawned automatically on server-workers in the Unreal Engine startup flow, so the user doesn't need to explicitly spawn these Singleton Actors.

Reviewer 3: ```suggestion
    - A caveat to the above rule is tha `AGameMode` and `AGameState` Singletons Actors are spawned automatically on server-workers as part of the Unreal Engine startup flow, so you don't need to explicitly spawn these Singleton Actors.
```
